### ISSUE TITLE
Events are not sent in AWS Lambda 

### ISSUE THREAD

**Issue by clouddra** [getsentry/sentry-dotnet#126]
I have discovered some issues when using SDK on AWS Lambda/API Gateway

The following message is logged to sentry  
```csharp
[HttpGet("/health")]
[ProducesResponseType((int) HttpStatusCode.OK)]
public async Task HealthCheck()
{
    _logger.LogCritical("test");
    await Task.Delay(TimeSpan.FromSeconds(10));
}
```

while this does not

```csharp
[HttpGet("/health")]
[ProducesResponseType((int) HttpStatusCode.OK)]
public IActionResult HealthCheck()
{
    _logger.LogCritical("test");
    return Ok()
}
```

The problem might be related to https://github.com/aws/aws-lambda-dotnet/issues/239. The thread is killed right after the response is sent.

**Comment by clouddra** [comment]
@bruno-garcia Thanks for the explanation. Would it be possible to integrate such blocking behavior directly into the sdk itself? Possibly a config flag that we can turn on for serverless environments, considering that such deployment modes are becoming more and more common.

**Comment by bruno-garcia** [comment]
Yes that makes sense. If the approach is something like I suggested above, perhaps it would fit in the `Sentry.AspNetCore` package since it only means implementing `IActionResult`. If somehow we require dependency on some AWS assembly, a new package would be created like `Sentry.Amazon.Aws` or something.
We'd be glad to merge a PR solving this one ;)

**Comment by clouddra** [comment]
Implementing `IActionResult` is does not seem ergonomic. Users can always return a plain object and bypassing the flush. Also, I presume calling `SentrySdk.Close()` would break subsequent requests in a long running server process (non serverless).

It be nicer if we are able to hook into the response lifecycle `Response.OnCompleted` in `SentryMiddleware` and call `Worker.Dispose()` to do trigger a manual flush without disposing `SentryClient`. However, there seemed to be no easy way to do this. Most of the useful APIs are hidden and not exposed in public interfaces.

**Comment by bruno-garcia** [comment]
@clouddra the APIs are hidden to allow evolving without necessarily breaking changes. We can make public things when they make sense and I'd be happy to merge a PR which adds the functionality in the SDK.

The suggestion to call `Close` at the end of the request would only make sense for serverless environments. The SDK is currently designed for long running processes so I believe that part is covered. Nevertheless I agree it doesn't make sense now that I read the lifecycle. @untitaker was dealing with this on the python SDK so I guess it's time we discuss this and agree on a solution.

Agreed the best way to go is the hook to `HttpContext.Response.OnCompleted`.

**Comment by untitaker** [comment]
I was proposing a `drain()` method right next to `SentryClient.close()` (or `SentrySdk.close()`, not sure where that method lives in C#) that, like `close()`, blocks until the queue is empty, but doesn't shut down the worker thread.

FWIW the common workaround in raven-python was to have a transport that just blocks calls to `captureException` and doesn't even have a worker thread. There are environments that would need this without alternative (https://github.com/getsentry/sentry-python/issues/125) so I wonder if we should add that back as well

**Comment by clouddra** [comment]
I tried playing around with the code and hooking `HttpContext.Response.OnCompleted` works. I tried to implement the `Drain()` API but there seems to be no easy way. To implement this neatly we will need to expose WorkerAsync instance and wait until queue is exhausted. There is no way to express `wait until queue is exhausted` since blocking semaphores prevents the `Task` from ever returning.

@bruno-garcia 
> One issue I see with drain trying to empty the queue is that you can have (I assume) multiple request on that same container all causing events which means all of those threads would be blocked on Drain() while new requests are coming in and queuing more events etc.

We can always remember the `last_event` in the queue and `Drain()` will then process event(s) up till `last_events`. New events added while `Drain()` is running will not be processed.

**Comment by bruno-garcia** [comment]
@clouddra 
> To implement this neatly we will need to expose WorkerAsync instance and wait until queue is exhausted. There is no way to express wait until queue is exhausted since blocking semaphores prevents the Task from ever returning.

That task only resolves once the worker finishes, which currently is triggered by `Dispose`. We'd need to have some signalling done between the `Drain()` call and the worker loop.
Considering the code is currently relying on `CancellationToken` to come back from awaiting the semaphore, it would need quite a bit of a re-write to support `Drain()`.

> We can always remember the last_event in the queue and Drain() will then process event(s) up till last_events. New events added while Drain() is running will not be processed.

That's the approach taken by @mitsuhiko originally on the Rust SDK before we decided to drop `drain`.

I guess we could use the `SentrySdk.LastEventId` to drain until the current scope's Id or else we'd need to implement some queue that supports `Peek` because that can't be done with `ConcurrentQueue`.

**Comment by untitaker** [comment]
>I guess we could use the SentrySdk.LastEventId to drain until the current scope's Id or else we'd need to implement some queue that supports Peek because that can't be done with ConcurrentQueue.

I would do this:

* get current length of queue, store it in `n`
* block until `n` events have been sent

I think this achieves the same thing

**Comment by clouddra** [comment]
@untitaker There is no need to block. The producer can still continue to push events since and the consumer consumes n messages.

**Comment by untitaker** [comment]
@clouddra that's not what I meant. In `Client.drain` you'd block until the next `n` events have been sent.

**Comment by bruno-garcia** [comment]
`SentrySdk` and `ISentryClient` have a `FlushAsync` method which we can use to make sure queued events are sent out. Ideally we'd hook somewhere to get called at the end of each request. Not sure if it's possible in AWS lambda though

**Comment by madmox** [comment]
Any update on this issue? We are currently hosting our Web API behind an AWS lambda, and this is the only thing preventing us from enabling sentry logs in our backend :/

**Comment by bruno-garcia** [comment]
Can you please elaborate how is this blocking you?

The current approach here is to call `await SentrySdk.FlushAsync()`. That'll flush any queued up event to Sentry.

The issue is still open so that we can look into finding a way to hook it nicely where the flush happens to your implicitly and no user code is required. Something like shipping a NuGet package `Sentry.Aws.Lambda` or whatever would be needed to magically flush pending events at the end of every lamda execution.

**Comment by madmox** [comment]
Maybe I just misinterpreted the above discussion. I thought multiple solutions where proposed and none of them was completely covering all of the use cases.

So if I understand correctly what you say, it's possible to write a [middleware](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/write?view=aspnetcore-2.1), which would flush the sentry events queue after each request by calling `FlushAsync` on an injected `ISentryClient` instance? Is this service even injectable by default?

Such as:
```csharp
public class SentryEventsFlusherMiddleware
{
    private readonly RequestDelegate _next;
    private readonly ISentryClient _sentryClient;

    public GitRevisionHashMiddleware(RequestDelegate next, ISentryClient sentryClient)
    {
        this._next = next;
        this._sentryClient = sentryClient;
    }

    public async Task InvokeAsync(HttpContext context)
    {
        await this._next(context);
        await this._sentryClient.FlushAsync();
    }
}
```
I'm not entirely sure how ASP.NET Core lambda integration interacts with the lambda execution context, the thread might be freezed right after `await this._next(context);` is called. Another option in this case would be:
```csharp
public class SentryEventsFlusherMiddleware
{
    private readonly RequestDelegate _next;
    private readonly ISentryClient _sentryClient;

    public GitRevisionHashMiddleware(RequestDelegate next, ISentryClient sentryClient)
    {
        this._next = next;
        this._sentryClient = sentryClient;
    }

    public async Task InvokeAsync(HttpContext context)
    {
        context.Response.OnStarting(async () => { await this._sentryClient.FlushAsync(); });
        await this._next(context);
    }
}
```

Also, note that we are hosting non-ASP.NET Core lambdas (but still .NET Core), which would benefit from Sentry logging (mainly SQS queues & SNS topics triggered lambdas). Given that what I proposed above works, the workaround here would be to inject an `ISentryClient` service dependency in the function class, and call `FlushAsync` right before each function returns. Right?

**Comment by bruno-garcia** [comment]
Your understanding in on spot. Also, the code you proposed should work fine. 
`ISentryClient` and `IHub` are both injectable by default.
I doubt lambda would freeze the process before the whole pipeline returns btw. I didn't test it myself but my guess is that it's safe to have as you exemplified.

If you are using `UseSentry`, your code is fine because Sentry SDK also adds the try/catch. 
If not, I suggest then

```csharp
try {
await _next(ctx);
} catch (Exception e) {
_sentry.CaptureException(e);
await _sentry.FlushAsync();
}
```

I'm unfamiliar with the other triggers in lambda that don't use ASP.NET Core.
My guess is that the easiest thing is to `Init` the SDK int he beginning of the function, add a try/catch and capture/flush events like above.

Finally, I don't think `context.Response.OnStarting(async () => { await this._sentryClient.FlushAsync(); });` is the way to go. You'd most likely want to flush once the response is completed. For one to not delay the response to the caller, and secondly to make sure you capture errors at any stage of the request.

**Comment by madmox** [comment]
> For one to not delay the response to the caller, and secondly to make sure you capture errors at any stage of the request.

Agreed with the "make sure you capture errors at any stage of the request".

But I'm not entirely sure the delay thing is avoidable. I suspect the ASP.NET Core lambda integration to return the data to the caller (using something like `return new APIGatewayProxyResponse(...)`) only after all ASP.NET Core events have fired. The `HttpContext.Response` might just me an in-memory proxy implementation feeding the `APIGatewayProxyResponse` to be returned after the ASP.NET MVC Core pipeline.

**Comment by bruno-garcia** [comment]
If that's the case it would be throwing away a lot of the benefits you get from using ASP.NET Core where the data would be written to the output stream as the application makes it available.

It would be interesting to measure that. Like for example send the response from a controller action, then "block" for a couple of seconds (await. `Task.Delay(2000)`) and send a test event to sentry (`_sentry.CaptureMessage($"took {stopwatch} long")`) so you can see on the browser if the response is coming before the 2 seconds, and the event in Sentry comes in after that.

**Comment by madmox** [comment]
In my understanding, an API gateway lambda function is a very simple piece of code:
```csharp
public class Function
{
    public async Task<APIGatewayProxyResponse> Get(APIGatewayProxyRequest request, ILambdaContext context)
    {
        // Your code goes here
        string data = "Hello, world!";
        
        return new APIGatewayProxyResponse() { StatusCode = 200, Body = data };
    }
}
```
The ASP.NET MVC Core integration is no exception to the rule (it's also the same thing with any other language, so python/node/etc. implementations should face the exact same issues). Ultimately, all the code it adds is inserted in the placeholder zone. Considering data starts being emitted to the client after the `return` statement is hit, I don't see how the integration would proceed to send data over the wire before that point. It's also the point where the lambda environment context freezes the executing lambda threads.

I'm basing my interpretation on [this answer](https://github.com/aws/aws-lambda-dotnet/issues/239#issuecomment-371683525) by @normj and the time I spent working with raw API gateway lambdas, but I might be wrong. Testing is indeed a good way to confirm the point :)

Note that this behavior is the one causing the original issue described by @clouddra , so it should not come as a surprise. In the end, serverless Web API main benefits are cost & scalability, but it also comes with a lot of drawbacks very difficult to anticipate.

**Comment by bruno-garcia** [comment]
@madmox That makes sense. Thanks for linking to the answer from @normj, testing it now would only confirm what he said but makes not much sense now since he's one of the authors of that SDK.

In that case the middleware at the root of the pipeline, calling `await FlushAsync` right after invoking the request delegate is the best we can do.
It'll make sure Sentry's SDK flushes any queued up event before we return to the API gateway.

**Comment by madmox** [comment]
The proposed workaround using middleware does not work in 1.2.0 because `ISentryClient` does not expose `FlushAsync` in this version... what do you suggest for this version? Or am I required to use 2.0.0-beta4?

```csharp
public class SentryEventsFlusherMiddleware
{
    private readonly RequestDelegate _next;

    public SentryEventsFlusherMiddleware(RequestDelegate next)
    {
        this._next = next;
    }

    public async Task InvokeAsync(HttpContext context, ISentryClient sentryClient)
    {
        context.Response.OnCompleted(async () =>
        {
            await sentryClient.FlushAsync(); // 'ISentryClient' does not contain a definition for 'FlushAsync'
        });

        await this._next(context);
    }
}
```

**Comment by bruno-garcia** [comment]
I'd suggest using 2.0.0-beta4.
The only reason it's not `2.0.0` (non preview) is that I want to fit some more "breaking" changes in it before but we're currently busy with a Android SDK so this is on hold.

**Comment by madmox** [comment]
Final note regarding the middleware: you should use `context.Response.OnCompleted` instead of flushing after `await this._next(context);`, otherwise you might miss any event logged after the middleware returns - e.g. middlewares higher in the ASP.NET pipeline, or any error logged by the lambda integration when marshalling the response to API gateway:

```csharp
public class SentryEventsFlusherMiddleware
{
    private readonly RequestDelegate _next;

    public SentryEventsFlusherMiddleware(RequestDelegate next)
    {
        this._next = next;
    }

    public async Task InvokeAsync(HttpContext context, ISentryClient sentryClient)
    {
        context.Response.OnCompleted(async () =>
        {
            await sentryClient.FlushAsync(timeout: TimeSpan.FromSeconds(10));
        });

        await this._next(context);
    }
}
```

For non-ASP.NET lambdas, I ended up writing a basic try/catch/finally block in the function:

```csharp
public class Function
{
    private readonly ILogger _logger;
    private readonly ISentryClient _sentryClient;

    public Function()
    {
        // Initialize dependency injection & dependencies here
    }

    public async Task ProcessSQSEvent(SQSEvent sqsEvent)
    {
        try
        {
            // Your code
        }
        catch (Exception ex)
        {
            this._logger.LogError(ex, ex.Message);
            throw; // Let AWS move SQS messages to deadletter
        }
        finally
        {
            await this._sentryClient.FlushAsync(timeout: TimeSpan.FromSeconds(10));
        }
    }
}
```

This all worked for me with 2.0.0-beta4.

**Comment by bruno-garcia** [comment]
@madmox Thanks for sharing the snippet.

>  Final note regarding the middleware: you should use context.Response.OnCompleted instead of flushing after await this._next(context);, otherwise you might miss any event logged after the middleware returns - e.g. middlewares higher in the ASP.NET pipeline, or any error logged by the lambda integration when marshalling the response to API gateway

That makes sense. I think the simplest way to add this to the SDK is through an option and conditionally flush after each request. I'll raise a ticket, maybe someone (you?) from the community would like to contribute this.

**Comment by normj** [comment]
Sounds like you all have reached consensus. I just wanted to chime that yes you have it right. Nothing is sent back to the client till the originating function has returned. In the ASP.NET Core Lambda case that is the `FunctionHandlerAsync` method in Amazon.Lambda.AspNetCoreServer that kicks off the ASP.NET Core pipeline and returns the results of the pipeline back to the client.

**Comment by bruno-garcia** [comment]
Thanks @normj for the details!
